source/extensions/quic_listeners/quiche/platform/flags_impl.cc

```diff
@@ -0,0 +1,102 @@
+// NOLINT(namespace-envoy)
+
+// This file is part of the QUICHE platform implementation, and is not to be
+// consumed or referenced directly by other Envoy code. It serves purely as a
+// porting layer for QUICHE.
+
+#include "extensions/quic_listeners/quiche/platform/flags_impl.h"
+
+#include <set>
+
+#include "absl/strings/ascii.h"
+#include "absl/strings/numbers.h"
+
+namespace quiche {
+
+namespace {
+
+absl::flat_hash_map<std::string, Flag*> MakeFlagMap() {
+  absl::flat_hash_map<std::string, Flag*> flags;
+
+#define QUICHE_FLAG(type, flag, value, help) flags.emplace(FLAGS_##flag->name(), FLAGS_##flag);
+#include "extensions/quic_listeners/quiche/platform/flags_list.h"
+#undef QUICHE_FLAG
+
+  return flags;
+}
+
+} // namespace
+
+// static
+FlagRegistry& FlagRegistry::GetInstance() {
+  static auto* instance = new FlagRegistry();
+  return *instance;
+}
+
+FlagRegistry::FlagRegistry() : flags_(MakeFlagMap()) {}
+
+void FlagRegistry::ResetFlags() const {
```

Comment:
Will QUIC tests work as expected if this isn't reset between test cases?

Reply:
Most QUIC tests won't work without the reset.
We have not enabled those tests in Envoy, because quic_flags_impl is not available yet:)

Reply:
QUICHE defines a [QuicFlagSaver](https://quiche.googlesource.com/quiche/+/89bfd47d351f2d014f7d9a8a2b431e0b57a4f69f/quic/platform/api/quic_test.h#11) utility to do this. Once this PR is in, I will fill in the platform impl for that (located [here](https://github.com/envoyproxy/envoy/blob/56f2f0d0161051a0b0eea1e1bad355a4668f1930/test/extensions/quic_listeners/quiche/platform/quic_test_impl.h#L17)) to call FlagRegistry::ResetFlags.  Any QUICHE tests that mess with flags should already be using QuicFlagSaver to reset their state.

Reply:
Gotcha.  Thanks for the context.
In that case I think you add a TODO for sorting out the Envoy runtime (just to make it clear we don't consider this a substitute and we plan to do it long term) and you're "good" to go, where "good" comes with a healthy dose of "I had no idea you could do such whacky things with includes and macros and I wish I still didn't" :-P

Reply:
Done. Thanks for the review!

I agree the macro-hackery is a bit, um, hackish :-)  I guess ideally, the QUICHE platform API would be a little less flag-centric and just a programmatic lookup API like Runtime::Snapshot. Then, the Envoy impl would be a direct pass-through, and google3/Chromium could implement with a flag backing.